### DISCUSSION TITLE
Cannot use injected Request with a method returning Handler

### DISCUSSION THREAD

**Discussion by Dkhusainov** [playframework/playframework#11566] [General]
### Play Version

2.8

### API

Java

### Expected Behavior

1. Create a route which accepts `play.mvc.Http.Request` and returns `play.api.mvc.Handler`
2. Compile and run it
3.

### Actual Behavior
Scala compilation fails with message `Cannot use a method returning play.mvc.Http.Request => play.api.mvc.Handler as a Handler for requests`

If I change return type to `CompletionStage`(for example) everything compiles as expected.

### Reproducible Test Case

```
POST         /request                                       com.Controller.request(req: Request)
```
```
public class Controller {
      public Handler request(Http.Request request) {
        return null;
    }
}
```

**Comment by mkurz** [comment]
Not sure why you think you can return a `Handler` from an action. You need to return a `Result`, please see the docs how to write controllers and actions: https://www.playframework.com/documentation/2.8.12/JavaActions
Request handlers are not used within controllers, you can read more about how to intercept http requests here: https://www.playframework.com/documentation/2.8.12/JavaActionCreator#HTTP-request-handlers

**Comment by Dkhusainov** [comment]
> Not sure why you think you can return a Handler from an action

But I can, and I do...

My controller methods generated and return `Handler`, which then invokes the actual implementation returning `CompletionStage<Result>`

It only breaks when I try to use Request as a parameter, hence the issue. Are  you saying I should just work around it myself, since this looks like an undocumented feature(using Handler directly)?

**Reply by mkurz** [reply]
Sorry for being late here, but why do you want to return a `Handler`, why not just use `Result` like documented? Where did you read about returning a Handler?
Just use the official way of the documentation and you should be ok.

**Comment by Dkhusainov** [comment]
Controller types which are injected into the router are generated(why they are generated is another topic). For each of those controllers an interface is also generated, which is implemented by my actual business logic controllers. The generated one injects the implementation one and invokes it's methods. 

So the generated ones return a `Handler` instead of a `Result` because they need to read annotations from business logic ones(`Action` and `BodyParser`), and pass it to returned `JavaAction`(aka `Handler`). Since I can't annotate a generated type, I have to do this.
